### ISSUE TITLE
Pre-processing help for Finetuning a model to NFCorpus.

### ISSUE THREAD

**Issue by Shaunak-Joshi** [UKPLab/sentence-transformers#2484]
### NFCORPUS
Medical dataset from [BeIR](https://github.com/beir-cellar/beir/tree/main). Each query has been mapped to a set of multiple passages. On average, there are 38.2 Relevant Documents per Query. 

### THE ISSUE
I wanted to finetune a model to this dataset. I am planning to finetune for improving Information Retrieval. I looked into the different evaluators and losses available. According to my understanding and following along with this [example](https://github.com/UKPLab/sentence-transformers/blob/master/examples/training/ms_marco/train_bi-encoder_mnrl.py), you need a one-to-one mapping of query and answer along with a 'score' to quantify how similar the two sentences are. So, to replicate this, I have coded the below script. I guess my question is does this make sense.

```
import faiss
import numpy as np
from sentence_transformers import SentenceTransformer, CrossEncoder

class Vectorstore:
  def __init__(self,model,documents):       #model is a Sentence Transformer Embedding Model
    self.index = faiss.IndexFlatL2(model.get_sentence_embedding_dimension())
    self.embeddings = model.encode(documents)
    faiss.normalize_L2(self.embeddings)
    self.index.add(self.embeddings)
    self.ce = CrossEncoder('cross-encoder/ms-marco-MiniLM-L-6-v2', device="cuda")

  def sigmoid(self,x):
    return 1/(1+np.exp(-x))

  def vectorStore(self,model,documents,query):
      tempDict = {}
      returnDict = {}

      search_vector = model.encode(query)
      _vector = np.array([search_vector])
      faiss.normalize_L2(_vector)
      distances, ann = self.index.search(_vector, k=10)

      for i,j in zip(distances[0],ann[0]):
        tempDict[documents[j]] = i

      for i in tempDict:
        returnDict[self.sigmoid(self.ce.predict([query,i]))] = i

      values = sorted(list(returnDict.keys()),reverse=True)[:2]

      answer = {}
      for i in values:
        answer[i] = returnDict[i]
      return answer
```

For every query, it iterates through all the relevant documents and returns a dict of `{ ce_score: relevant_doc }`. 

I am planning to [Cosine Similarity Loss](https://github.com/UKPLab/sentence-transformers/blob/master/sentence_transformers/losses/CosineSimilarityLoss.py) and [Embedding Similarity Evaluator](https://github.com/UKPLab/sentence-transformers/blob/master/sentence_transformers/evaluation/EmbeddingSimilarityEvaluator.py) for training.

**Comment by Shaunak-Joshi** [comment]
@ir2718 

Hello, thank you for the reply. 

I didn't understand what you meant by adding a Bi-Encoder. I am using a Bi-Encoder originally to 

1. Embed all the sentences and store in Faiss 
2. Use Faiss Index and query it. It returns a set of 10 documents. 

Then I pass, the retrieved 10 sentences to the Cross Encoder one-by-one to get a score of one-to-one similarity of the query and sentence. This score is then normalized using sigmoid to 0 to 1. These documents are re-ranked using the score and the top most pair is essentially being planned to be used as training data. 

So,
Embed documents using Bi-Encoder --> Create Index --> Query Index and return set of 10 documents --> Re-rank using Cross Encoder --> Return top most (query, sentence, CE Score)

The score is acting as the label float similarity score for [Cosine Similarity Loss](https://github.com/UKPLab/sentence-transformers/blob/master/sentence_transformers/losses/CosineSimilarityLoss.py). Creating (query, sentence, Cross-Encoder Normalized Score). 

I didn't understand where in the flow should I insert a Bi-Encoder. It is already being used to compute embeddings, Cross Encoder is only being used to "re-rank" the **10** query-document pairs. 

Thanks!

**Comment by ir2718** [comment]
Okay, I was missing the part were you use the bi-encoder to create the initial embeddings, and then you do re-ranking using a cross-encoder. If this is the way you do it then disregard my previous comment. The approach you described seems sound and should work.

**Comment by Shaunak-Joshi** [comment]
No worries. 

One more thing, I trained the model using Cosine Similarity Loss and Embedding Similarity Evaluator. For 10 epochs. But the accuracy steadily plummeted after 10. I recreated this for 6 models and all of them followed the same trend. Is there a reason for this? 

Thanks.

**Comment by tomaarsen** [comment]
Hello!

> But the accuracy steadily plummeted after 10.

What do you mean with "after 10"? Do you mean after training? Or do you mean if you continue training after the 10 epochs?

---

Some other suggestions:
1. The extremely new `CoSENTLoss` may outperform `CosineSimilarityLoss` while having the same usage (consider it a drop-in replacement). You can get access to it by installing the bleeding edge version of Sentence Transformers: `pip install git+https://github.com/UKPLab/sentence-transformers.git`.
2. `MultipleNegativesRankingLoss` is one of the strongest loss functions currently. It works by providing it with `InputExample` instances of *positive pairs*, i.e. a query and a passage. You can preprocess your dataset very easily to work with these positive pairs: simply include all query-document pairs. It is recommended to use this loss in conjunction with the [NoDuplicatesDataLoader](https://www.sbert.net/docs/package_reference/datasets.html#noduplicatesdataloader). This ensures that there are no duplicate sentences in a batch, which is important for MNRL as it crucially considers all other sentences in the batch as "negative pairs", i.e. texts that are unrelated to the query. Sidenote: higher batch size => (generally) better performance for MNRL.

I would experiment with these two, and optionally, even use both simultaneously, e.g. like [here](https://www.sbert.net/examples/training/quora_duplicate_questions/README.html#multi-task-learning). If you'd then like to get even more performance out of your model, then you can also experiment with [CachedMultipleNegativesRankingLoss](https://www.sbert.net/docs/package_reference/losses.html#cachedmultiplenegativesrankingloss) instead of MNRL. It is like MNRL except you can train with massive batch sizes (i.e. better performance, generally) without needing a ton of GPU memory.

- Tom Aarsen

**Comment by Shaunak-Joshi** [comment]
@tomaarsen 

Hello! Thanks for the reply!

> What do you mean with "after 10"? Do you mean after training? Or do you mean if you continue training after the 10 epochs?

I meant after I continue training for 10+ epochs (20, 30), the accuracy consistently dropped. I saw this trend for multiple models, that is why I wondered about this specifically. 

> The extremely new `CoSENTLoss` may outperform `CosineSimilarityLoss`

I will look into this as I have barely explored this loss. Thanks

>crucially considers all other sentences in the batch as "negative pairs"

I must have completely missed this when reading the documentation or interpreted it wrong. This is new information for me. 

I'll explore the resources you have mentioned. Thank you so much!

**Comment by tomaarsen** [comment]
Hello!

> I meant after I continue training for 10+ epochs (20, 30), the accuracy consistently dropped. I saw this trend for multiple models, that is why I wondered about this specifically.

Most NLP models are trained using specific learning rate schedulers, which slowly decrease the learning rate from the starting value down to 0. So, if you train a model for 10 epochs, it will be very slowly making final modifications when nearing epoch 10. If you then load that model and train some more, it will once again train using much higher learning rates, which can heavily reduce the accuracy right at the start of training. 

So, if you'd like to experiment with training a model for 20 or 30 epochs, then I would recommend training the "untrained" base model for the full 20 or 30 epochs rather than continuing training on the model that was already trained for 10 epochs. It will likely perform better.

> > crucially considers all other sentences in the batch as "negative pairs"
>
> I must have completely missed this when reading the documentation or interpreted it wrong. This is new information for me.

To clarify a bit further: This is how the `MultipleNegativesRankingLoss` works - this does not apply for other losses. I'm not sure if I made that clear in my comment :)

- Tom Aarsen

**Comment by Shaunak-Joshi** [comment]
Hi!

> So, if you train a model for 10 epochs, it will be very slowly making final modifications when nearing epoch 10.

This makes a lot of sense. I was using [SBERT Pre-Trained Models](https://www.sbert.net/docs/pretrained_models.html) for finetuning as a I figured that for some basic exploration that makes a lot of sense. But, finetuning the base model is much more logical now that I think about it.

Regarding MNRL loss, I always figured it needed both positives as well as "hard-negatives" as an Input. But now I have the option to explore MNRL as well. I didn't have a way to create a hard negative in the original code of this issue, that is why I avoided MNRL.

Regardless, thank you for the clarifications as well as the resources you have provided. Amazing work on this repo as always!

Thank you!
